File: cpp/Attribute.cpp
//
// Created by Dylan on 4/27/2021.
//

#include <map>
#include <vector>
#include <string>
#include "Attribute.h"

class Attribute;

class Attribute {
    public:
        std::string name;
        weak_ptr<Organism> organism;

        Attribute(std::string name, weak_ptr<Organism> organism) {
            this->name = name;
            this->organism = organism;
        }

        virtual AttributeOutput eval(const EvalContext& evalContext) = 0;
        virtual ~Attribute() {}
};

File: js/go/attribute.go
package main

type Attribute struct {
	RootNode Node
	Name
	Id
	onRootNodeChanged chan struct{}
}

func NewAttribute() *Attribute {
	return &Attribute{onRootNodeChanged: make(chan struct{})}
}

func (a Attribute) eval() float32 {
	return a.RootNode.eval()
}

func (a *Attribute) setRootNode(node Node) {
	node.setAttribute(a)
	a.RootNode = node
	go func() {
		a.onRootNodeChanged <- struct{}{}
	}()
}
